### PR TITLE
WT-8264 Add a CMake build support for workgen

### PR DESCRIPTION
**Pull request by alisonfel**
Currently, we're missing support for building workgen with CMake. This was missed in the initial pass when CMake support was introduced. This ticket will introduce a CMake build script to support compiling workgen. Specifically this PR adds the following changes:
* A CMakeLists.txt script to compile the workgen sources. Similar to the Python API, the script executes SWIG to generate the workgen wrapper python sources.
* Refactored out the Python package & install sourcing as a general helper macro. We need this common code to build our Python workgen library.
***
_This is one of a series of CMake changes needed to complete WT-8020._

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -195,3 +195,4 @@ enable_testing()
 add_subdirectory(test)
 add_subdirectory(examples)
 add_subdirectory(bench/wtperf)
+add_subdirectory(bench/workgen)
```
Comment (etienneptl) [inline CMakeLists.txt:198]: Is it possible to alphabetize this list or is there a specific order to respect?

File: bench/workgen/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+set(python_libs)
+set(python_version)
+set(python_executable)
+# Source the users Python3 install, extracting the Python 3 version, library and interpreter install paths.
+source_python3_package(python_libs python_version python_executable)
+
+# Compile the intermediate C++ workgen sources into static library.
+add_library(workgen_cxx
+    STATIC
```
Comment (etienneptl) [inline bench/workgen/CMakeLists.txt:70]: format np: you probably want `workgen_cxx` above `STATIC` like you did lines 75/76

**Review by etienneptl — APPROVED** [review]
LGTM

File: CMakeLists.txt (hunk 1 above)
Comment (alisonfel) [reply CMakeLists.txt:195]: In lieu of of the greater restyling ticket, I haven't been really enforcing that convention in the CMake sources. Though I've applied it here in this specific case `88652cb `

File: bench/workgen/CMakeLists.txt (hunk 2 above)
Comment (alisonfel) [reply bench/workgen/CMakeLists.txt:71]: Restyled.

File: bench/workgen/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+list(APPEND swig_flags "-I${CMAKE_BINARY_DIR}/include")
+list(APPEND swig_flags "-I${CMAKE_BINARY_DIR}/config")
+
+set(CMAKE_SWIG_FLAGS ${swig_flags})
+# By default, the CMake SWIG module expects the swig interface file extension to be ".i". Newer
+# versions of CMake 3.14+ support specifying 'SWIG_SOURCE_FILE_EXTENSIONS' to consider different
+# extensions. Though given our minimum support version is 3.10, we need to copy the file (and its header dependency)
+# over to our binary directory using the ".i" extension.
+configure_file(${CMAKE_CURRENT_LIST_DIR}/workgen.h ${CMAKE_CURRENT_BINARY_DIR}/workgen.h COPYONLY)
+configure_file(${CMAKE_CURRENT_LIST_DIR}/workgen.swig ${CMAKE_CURRENT_BINARY_DIR}/workgen.i COPYONLY)
```
Comment (ddanderson) [inline bench/workgen/CMakeLists.txt:110]: If it's more convenient or standard, you could rename `workgen.swig` to be `workgen.i` .  I used `.swig` to distinguish from C include files that were named with `.i`, but include files have been renamed to be `_inline.h`.  Though `.swig` may be more readable for anyone casually browsing directories.

File: bench/workgen/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    set(wiredtiger_target wiredtiger_shared)
+else()
+    message(STATIC "Skipping workgen compiliation: Requires either a shared library (ENABLE_SHARED)"
+                    "or a PIC-enabled (WITH_PIC) build of wiredtiger")
+endif()
+
+set(python_libs)
+set(python_version)
+set(python_executable)
+# Source the users Python3 install, extracting the Python 3 version, library and interpreter install paths.
```
Comment (ddanderson) [inline bench/workgen/CMakeLists.txt:71]: `users` -> `user's`

**Review by ddanderson — APPROVED** [review]
LGTM, a couple small comments.

File: bench/workgen/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+set_property(TARGET workgen_cxx PROPERTY POSITION_INDEPENDENT_CODE ON)
+
+# Manually specify the interface name '_workgen' over a generated name.
+# Our workgen tooling expects the swig interface name '_workgen'
+# when importing the generated swig library.
+list(APPEND swig_flags "-c++")
+list(APPEND swig_flags "-interface;_workgen")
+list(APPEND swig_flags "-threads")
+list(APPEND swig_flags "-O")
+list(APPEND swig_flags "-Wall")
```
Comment (jiechenbo) [inline bench/workgen/CMakeLists.txt:99]: I'm not sure how autoconf configures the CPP flags, but is it possible to have same diagnostic CPP warnings flags from the cppsuite, be implemented here too?

**Review by jiechenbo — APPROVED** [review]
LGTM! I have just one question.

File: bench/workgen/CMakeLists.txt (hunk 5 above)
Comment (etienneptl) [reply bench/workgen/CMakeLists.txt:99]: I had the same question as you @jiechenbo. I think we will add the cxx flags to the workgen suite if we want to, but it's definitely out of scope here and we may have a ticket for it already!

Comment (alisonfel) [reply bench/workgen/CMakeLists.txt:99]: Due to the auto-generated nature of the SWIG wrapper, a lot of our `strict` diagnostic warnings are incompatible and noisy. Any diagnostic flags we want to apply need to be tailored to workgen's compilation (which is similar to how we build the Python API) and hence is out of the scope of this ticket. 

As a side, the line this comment is set on are `swig` compiler flags (not `g++`/`clang` flags). Such flags would be set below in `swig_cxx_flags`.

Comment (etienneptl) [reply bench/workgen/CMakeLists.txt:99]: @alisonfel, thanks for the clarification!

File: bench/workgen/CMakeLists.txt (hunk 3 above)
Comment (alisonfel) [reply bench/workgen/CMakeLists.txt]: Thanks for background and clarification! Using  `workgen.i` would make it much more convenient for CMake (both to meet its expectations that SWIG interface files conventionally use the `.i` extension and to track the original source  as a file-level dependency). I've updated the PR, renaming the file to `workgen.i`.

File: bench/workgen/CMakeLists.txt (hunk 4 above)
Comment (alisonfel) [reply bench/workgen/CMakeLists.txt]: Good find. Fixed :)

File: bench/workgen/CMakeLists.txt (hunk 3 above)
Comment (ddanderson) [reply bench/workgen/CMakeLists.txt]: @alisonfel, just checked your change for `.swig` -> `.i`, looks good.

**Pull request merged**